### PR TITLE
ScopeTimer

### PR DESCRIPTION
**Pull request by JoshStoddard**
Adds a new utility `ScopeTimer` that measures the duration of its lifecycle using RAII.

e.g.:
```cpp
void critical_operation() {
    ScopeTimer<> _timer([](double elapsed_sec){
        std::cout << "critical_operation() took " << elapsed_sec << " seconds." << std::endl;
    });
    // do stuff
}
```

### CODE REVIEW

File: include/metrics/Timer.h [hunk 1]
```diff
@@ -23,28 +22,20 @@ template<
...
+    using timer_type = ScopeTimer< TIME_UNIT, double >;
+    using duration_type = typename timer_type::duration_type;
 
 private:
-    metric_type* _timer;
-    timespec _start;
+    std::unique_ptr< timer_type > _timer;
 
 public:
     explicit Timer(metric_type* timer = nullptr,
```
Comment (drewaogle) [inline include/metrics/Timer.h:34]: timer being prometheus::Histogram and _time being ScopeTimer is a bit confusing for review, at least.

File: include/metrics/Timer.h [hunk 2]
```diff
@@ -53,34 +44,28 @@ class Timer
...
         prometheus::Counter* start_counter = nullptr)
     {
-        timespec now;
-        clock_gettime(CLOCK_MONOTONIC, &now);
-        if (_timer) {
-            duration_type dur =
-                std::chrono::nanoseconds(now.tv_nsec - _start.tv_nsec) +
-                std::chrono::seconds(now.tv_sec - _start.tv_sec);
-            _timer->Observe(dur.count());
+        if (metric) {
```
Comment (drewaogle) [inline include/metrics/Timer.h:61]: so what this does here is create the ScopeTimer, and creates a function which on the ScopeTimer _timer destructor pushes the value out and into metric?

File: include/util/ScopeTimer.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    using callback_type = std::function< void(TIME_REP) >;
+    using duration_type = typename std::chrono::duration< TIME_REP, typename TIME_UNIT::period >;
+
+private:
+    callback_type _cb;
+    timespec _start;
+
+public:
+    template< typename CALLBACK >
+    explicit ScopeTimer(CALLBACK cb)
```
Comment (drewaogle) [inline include/util/ScopeTimer.h:36]: can this not be callback_type?

File: include/metrics/Timer.h (hunk 2 above)
Comment (JoshStoddard) [reply include/metrics/Timer.h:61]: yes, exactly.  In other words, creates an object that when it is destructed will automatically record the duration of its existence to `metric`.

Comment (drewaogle) [reply include/metrics/Timer.h:61]: nice, makes the metrics timer very clean

File: include/util/ScopeTimer.h (hunk 3 above)
Comment (drewaogle) [reply include/util/ScopeTimer.h]: looks great

File: include/metrics/Timer.h (hunk 1 above)
Comment (drewaogle) [reply include/metrics/Timer.h]: perfect!

**Review by drewaogle — APPROVED** [review]

**Pull request merged**